AC.andNot(RC)

ppiotrow opened the pull request:
I've decided to try using `Util.advanceUntil` in `andNot` between `ArrayContainer` and `RunContainer`. The implementation present in the commit was verified with present tests and a new one, but for sure I'd test it more if going to be merged.

I run two benchmarks `part{3,4}_andNotArrayContainerVSRunContainerContainer` on `master` and the new branch.
```
BasicAndNotContainerBenchmark.part3_andNotArrayContainerVSRunContainerContainer  avgt   15  3327,444 ± 49,353  ns/op
BasicAndNotContainerBenchmark.part4_andNotArrayContainerVSRunContainerContainer  avgt   15  1877,370 ± 25,888  ns/op

BasicAndNotContainerBenchmark.part3_andNotArrayContainerVSRunContainerContainer  avgt   15  7790,214 ± 81,502  ns/op
BasicAndNotContainerBenchmark.part4_andNotArrayContainerVSRunContainerContainer  avgt   15  638,899 ± 4,385  ns/op
```

The code is slower for the present benchmark which `andNot` Array Container of 1024 elements with very dense Run Container of 1009 runs with 64512 cardinality.
The new benchmark uses Run Container with 5 very long runs and the new code is 3x faster.

I'm opening this Pull Request to discuss
- is it worth to use new approach?
- should there be a switch when to use old and new approach?

I think the new approach will be faster when there are long sequences of array elements covered by runs or between two runs.
If there are very few array elements per run or between runs the old approach should be faster.

richardstartin commented:
Looks good to me but might it be worth using Thoughput and Sample bench mark modes since they paint a fuller picture? A Sample benchmark might have pointed to the speculative allocation of the short[] in the old code.

lemire commented:
Any chance we might dynamically detect which code path is fastest?

ppiotrow commented:
Richard, sure there will be more benchmarking. Not really understand "speculative allocation of the short[] as it occurs in both versions.

The bug is detected and I can fix it with next commit. As I marked the code is draft and requires to be properly tested.

Daniel, we may try to estimate how many array elements occur per a Run.
```java
int k = 10; //arbitraryEmelentsPerRun  - todo requires fine-tuning
if ((ac.cardinality/rc.numberOfRuns)>k) newCode() else oldCode()
```

lemire commented:
Yes, so I suggest you look at such a heuristic to choose the fast code path quicky.

richardstartin commented:
@ppiotrow 

On master I see unnecessary allocation for empty run containers:

```
233     short[] buffer = new short[cardinality]; 
234     if (x.nbrruns == 0) { 
235       return clone(); 
236     } 
```

lemire commented:
@ppiotrow  Can you update us on the status of this PR?

ppiotrow commented:
Sure, sorry for letting it hang without progress. I spent some time benchmarking and it actually showed that the improvement is not because of "smart search" by `advanceUntil` but by iterating over "runs" . I prepared method `iterateUntil` which - you know - continuously iterates ;) It appeared that it could almost replace `advanceUntil` even when the benchmark data was prepared to give boost for non-sequential search. I'll get back to it but recently I need to deal with non-development issues.

lemire commented:
@ppiotrow Any update on this issue?

ppiotrow commented:
@lemire Summer time slowly ends here and I've get back to that. I've noticed that there are no changes in performance which is caused by #178 .

lemire commented:
Ok. Let us see if we can fix https://github.com/RoaringBitmap/RoaringBitmap/issues/178

ppiotrow commented:
I've pushed rebased code to publish the recent changes. I've decided for now to switch to the `iterateUntil` as I can't clearly seem much improvement even in artificial benchmarks.

I'm looking for ideas how to benchmark the code. I've been using `RealDataBenchmarkAndNot.pairwiseAndNot` but it was not showing a difference even between hardcoded `byCardinalityAndNot` and `byRunsAndNot`.
The current code is fastest in both `part3` and `part4` version of `_andNotArrayContainerVSRunContainerContainer`, but the feature switch should be tuned/adjusted.

ppiotrow commented on src/main/java/org/roaringbitmap/ArrayContainer.java:
```diff
@@ -227,18 +227,25 @@ public ArrayContainer andNot(BitmapContainer value2) {
+    if (x.numberOfRuns() == 0) {
+      return clone();
+    }
+    final boolean byCardinality = (cardinality / x.numberOfRuns()) < 100;
```
this "feature switch" requires tuning

lemire commented:
@ppiotrow Now that we appear to have solved https://github.com/RoaringBitmap/RoaringBitmap/issues/178 Can you have a look again?

ppiotrow commented:
I'd like to get back to it again in January. I think that fixing #178 might be helpful, as I was not able to spot improvements in benchmarks where there should be some.

lemire commented:
@ppiotrow Ping!

ppiotrow commented:
I've force pushed the changes to this branch. They include the safe changes that could be merged now:
- removed speculative allocation of array containers
- andNot with "full" run container case
- removed need to decrement `i`
- various tests

I tried to benchmark the "by-runs" approach but still no luck (including the benchmark that extracts only array&run containers to `andNot` them).
If you accept the simple changes I'll port them to buffer implementation and we close it for now.

lemire commented:
Great! Let us review!

ppiotrow commented:
Change of plans - I've just pushed the real data benchmarks and switched by default (no feature switch) to the `by-runs` implementation, which has better results. The results are for the newest code, I've just manually switched to one of by-runs or by-cardinality approach. I've run them on the official benchmark server via `./run.sh org.roaringbitmap.runcontainer.ArrayContainerAndNotRunContainerBenchmark` 

### by-cardinality only
```
Benchmark                                                                  (dataset)   Mode  Cnt      Score     Error  Units
ArrayContainerAndNotRunContainerBenchmark.pairwiseACAndNotRC           census-income  thrpt    6   4241.531 ±  15.334  ops/s
ArrayContainerAndNotRunContainerBenchmark.pairwiseACAndNotRC              census1881  thrpt    6   3778.507 ±   7.461  ops/s
ArrayContainerAndNotRunContainerBenchmark.pairwiseACAndNotRC           dimension_008  thrpt    6    825.106 ±   6.128  ops/s
ArrayContainerAndNotRunContainerBenchmark.pairwiseACAndNotRC           dimension_003  thrpt    6    540.858 ±   2.518  ops/s
ArrayContainerAndNotRunContainerBenchmark.pairwiseACAndNotRC           dimension_033  thrpt    6    793.983 ±   3.601  ops/s
ArrayContainerAndNotRunContainerBenchmark.pairwiseACAndNotRC            uscensus2000  thrpt    6  26027.946 ± 192.707  ops/s
ArrayContainerAndNotRunContainerBenchmark.pairwiseACAndNotRC         weather_sept_85  thrpt    6   1519.658 ±  13.538  ops/s
ArrayContainerAndNotRunContainerBenchmark.pairwiseACAndNotRC      wikileaks-noquotes  thrpt    6   1974.499 ±  21.150  ops/s
ArrayContainerAndNotRunContainerBenchmark.pairwiseACAndNotRC       census-income_srt  thrpt    6   1474.671 ±   1.775  ops/s
ArrayContainerAndNotRunContainerBenchmark.pairwiseACAndNotRC          census1881_srt  thrpt    6  10918.254 ± 118.457  ops/s
ArrayContainerAndNotRunContainerBenchmark.pairwiseACAndNotRC     weather_sept_85_srt  thrpt    6    975.964 ±   2.285  ops/s
ArrayContainerAndNotRunContainerBenchmark.pairwiseACAndNotRC  wikileaks-noquotes_srt  thrpt    6   5456.867 ±  30.329  ops/s
```
### by-runs  only
```
Benchmark                                                                  (dataset)   Mode  Cnt      Score    Error  Units
ArrayContainerAndNotRunContainerBenchmark.pairwiseACAndNotRC           census-income  thrpt    6   4107.709 ±  9.203  ops/s
ArrayContainerAndNotRunContainerBenchmark.pairwiseACAndNotRC              census1881  thrpt    6   4015.516 ± 57.414  ops/s
ArrayContainerAndNotRunContainerBenchmark.pairwiseACAndNotRC           dimension_008  thrpt    6   1596.529 ± 39.238  ops/s
ArrayContainerAndNotRunContainerBenchmark.pairwiseACAndNotRC           dimension_003  thrpt    6    546.158 ±  1.116  ops/s
ArrayContainerAndNotRunContainerBenchmark.pairwiseACAndNotRC           dimension_033  thrpt    6   2211.556 ± 17.756  ops/s
ArrayContainerAndNotRunContainerBenchmark.pairwiseACAndNotRC            uscensus2000  thrpt    6  26200.215 ± 29.353  ops/s
ArrayContainerAndNotRunContainerBenchmark.pairwiseACAndNotRC         weather_sept_85  thrpt    6   1573.653 ± 15.746  ops/s
ArrayContainerAndNotRunContainerBenchmark.pairwiseACAndNotRC      wikileaks-noquotes  thrpt    6   2921.767 ±  4.562  ops/s
ArrayContainerAndNotRunContainerBenchmark.pairwiseACAndNotRC       census-income_srt  thrpt    6   2430.718 ±  6.003  ops/s
ArrayContainerAndNotRunContainerBenchmark.pairwiseACAndNotRC          census1881_srt  thrpt    6  11145.742 ± 68.725  ops/s
ArrayContainerAndNotRunContainerBenchmark.pairwiseACAndNotRC     weather_sept_85_srt  thrpt    6   1241.319 ±  4.702  ops/s
ArrayContainerAndNotRunContainerBenchmark.pairwiseACAndNotRC  wikileaks-noquotes_srt  thrpt    6   7610.497 ± 91.853  ops/s
```

In all benchmarks the by-runs approach has better throughput, especially `dimension_008`,`dimension_033` and `wikileaks-noquotes`. No improvement in ` uscensus2000` is because this is dead benchmark, no `andNot` is actually performed (could be explained by non overlapping containers etc).

I'm not using the `advanceUntil` binary search, the speed gain should be caused by better writes (`System.arrayCopy`). Instead of that I'm using `iterateUntil` that has different contract unfortunately
. I decided fit it *exactly* to my needs. Might be necessary to  remove it from very general class `Utils`.

lemire commented:
The CI fails...

```
Tests in error: 
  TestUtil.testIterateUntil:96 » ArrayIndexOutOfBounds -1
```

ppiotrow commented:
Right now it only fails on test coverage decreased. This is because there is a `byCardinality` method that is never used. It will be removed once we agree to always use `byRuns`.

lemire commented:
Ok. So your experimental results are strongly positive and the new code is small and understandable. You also pass our tests.

Any good reason to merge the now dead code?

ppiotrow commented:
No reason to leave old code. I'll remove it while implementing the same for the buffered version.

lemire commented:
@ppiotrow  Note that you must re-benchmark for the buffer case. It happens regularly that optimizations that work for the regular array-based code fails in the bytebuffer approach due to the different costs involved.

ppiotrow commented:
I've rebased once again as there were conflicts. The latest commits migrated the new `andNot` to the `Mappeable` containers.
Benchmark results:
### before
```
ArrayContainerAndNotRunContainerBenchmark.immutablePairwiseACAndNotRC           census-income  thrpt    6   4115.511 ±  23.200  ops/s
ArrayContainerAndNotRunContainerBenchmark.immutablePairwiseACAndNotRC              census1881  thrpt    6   3700.978 ±  18.233  ops/s
ArrayContainerAndNotRunContainerBenchmark.immutablePairwiseACAndNotRC           dimension_008  thrpt    6    715.214 ±   2.405  ops/s
ArrayContainerAndNotRunContainerBenchmark.immutablePairwiseACAndNotRC           dimension_003  thrpt    6    392.583 ±   2.620  ops/s
ArrayContainerAndNotRunContainerBenchmark.immutablePairwiseACAndNotRC           dimension_033  thrpt    6    957.163 ±   1.972  ops/s
ArrayContainerAndNotRunContainerBenchmark.immutablePairwiseACAndNotRC            uscensus2000  thrpt    6  13781.241 ±  23.237  ops/s
ArrayContainerAndNotRunContainerBenchmark.immutablePairwiseACAndNotRC         weather_sept_85  thrpt    6   1542.546 ±  21.425  ops/s
ArrayContainerAndNotRunContainerBenchmark.immutablePairwiseACAndNotRC      wikileaks-noquotes  thrpt    6   1918.950 ±  13.967  ops/s
ArrayContainerAndNotRunContainerBenchmark.immutablePairwiseACAndNotRC       census-income_srt  thrpt    6   1633.116 ±   4.996  ops/s
ArrayContainerAndNotRunContainerBenchmark.immutablePairwiseACAndNotRC          census1881_srt  thrpt    6   7220.821 ± 118.992  ops/s
ArrayContainerAndNotRunContainerBenchmark.immutablePairwiseACAndNotRC     weather_sept_85_srt  thrpt    6    856.222 ±   5.260  ops/s
ArrayContainerAndNotRunContainerBenchmark.immutablePairwiseACAndNotRC  wikileaks-noquotes_srt  thrpt    6   4956.840 ±   6.646  ops/s
```
### after
```
ArrayContainerAndNotRunContainerBenchmark.immutablePairwiseACAndNotRC           census-income  thrpt    6   3003.861 ± 11.617  ops/s
ArrayContainerAndNotRunContainerBenchmark.immutablePairwiseACAndNotRC              census1881  thrpt    6   3879.833 ± 23.042  ops/s
ArrayContainerAndNotRunContainerBenchmark.immutablePairwiseACAndNotRC           dimension_008  thrpt    6   1138.537 ± 12.382  ops/s
ArrayContainerAndNotRunContainerBenchmark.immutablePairwiseACAndNotRC           dimension_003  thrpt    6    424.930 ±  3.149  ops/s
ArrayContainerAndNotRunContainerBenchmark.immutablePairwiseACAndNotRC           dimension_033  thrpt    6   1964.210 ±  7.404  ops/s
ArrayContainerAndNotRunContainerBenchmark.immutablePairwiseACAndNotRC            uscensus2000  thrpt    6  13825.639 ± 61.090  ops/s
ArrayContainerAndNotRunContainerBenchmark.immutablePairwiseACAndNotRC         weather_sept_85  thrpt    6   1499.531 ± 10.418  ops/s
ArrayContainerAndNotRunContainerBenchmark.immutablePairwiseACAndNotRC      wikileaks-noquotes  thrpt    6   2041.218 ±  3.736  ops/s
ArrayContainerAndNotRunContainerBenchmark.immutablePairwiseACAndNotRC       census-income_srt  thrpt    6   1617.040 ±  5.009  ops/s
ArrayContainerAndNotRunContainerBenchmark.immutablePairwiseACAndNotRC          census1881_srt  thrpt    6   7130.179 ± 82.599  ops/s
ArrayContainerAndNotRunContainerBenchmark.immutablePairwiseACAndNotRC     weather_sept_85_srt  thrpt    6    876.837 ±  6.255  ops/s
ArrayContainerAndNotRunContainerBenchmark.immutablePairwiseACAndNotRC  wikileaks-noquotes_srt  thrpt    6   5662.352 ± 42.143  ops/s
```

I can see in results that some of them are faster while other slower. One more thing is that the benchmarks do not test off-heap. Have we ever tried the benchmarks for off heap?

ppiotrow commented:
Last commit fixes double allocation of `short[]` and `ShortBuffer`. It gives very good performance gain.
```
ArrayContainerAndNotRunContainerBenchmark.immutablePairwiseACAndNotRC           census-income  thrpt    6   3109.274 ±   8.578  ops/s
ArrayContainerAndNotRunContainerBenchmark.immutablePairwiseACAndNotRC              census1881  thrpt    6   3922.203 ±  25.543  ops/s
ArrayContainerAndNotRunContainerBenchmark.immutablePairwiseACAndNotRC           dimension_008  thrpt    6   1412.179 ±   4.474  ops/s
ArrayContainerAndNotRunContainerBenchmark.immutablePairwiseACAndNotRC           dimension_003  thrpt    6    428.023 ±   3.425  ops/s
ArrayContainerAndNotRunContainerBenchmark.immutablePairwiseACAndNotRC           dimension_033  thrpt    6   2251.272 ±  64.407  ops/s
ArrayContainerAndNotRunContainerBenchmark.immutablePairwiseACAndNotRC            uscensus2000  thrpt    6  13810.724 ±  42.388  ops/s
ArrayContainerAndNotRunContainerBenchmark.immutablePairwiseACAndNotRC         weather_sept_85  thrpt    6   1449.482 ±   4.666  ops/s
ArrayContainerAndNotRunContainerBenchmark.immutablePairwiseACAndNotRC      wikileaks-noquotes  thrpt    6   2370.417 ±  20.323  ops/s
ArrayContainerAndNotRunContainerBenchmark.immutablePairwiseACAndNotRC       census-income_srt  thrpt    6   1841.769 ±   6.374  ops/s
ArrayContainerAndNotRunContainerBenchmark.immutablePairwiseACAndNotRC          census1881_srt  thrpt    6   7177.638 ± 124.841  ops/s
ArrayContainerAndNotRunContainerBenchmark.immutablePairwiseACAndNotRC     weather_sept_85_srt  thrpt    6   1031.121 ±   5.117  ops/s
ArrayContainerAndNotRunContainerBenchmark.immutablePairwiseACAndNotRC  wikileaks-noquotes_srt  thrpt    6   6420.377 ±  78.892  ops/s
ArrayContainerAndNotRunContainerBenchmark.pairwiseACAndNotRC                    census-income  thrpt    6   3872.256 ±   8.075  ops/s
ArrayContainerAndNotRunContainerBenchmark.pairwiseACAndNotRC                       census1881  thrpt    6   4135.116 ±  46.413  ops/s
ArrayContainerAndNotRunContainerBenchmark.pairwiseACAndNotRC                    dimension_008  thrpt    6   1955.400 ±  22.059  ops/s
ArrayContainerAndNotRunContainerBenchmark.pairwiseACAndNotRC                    dimension_003  thrpt    6    568.534 ±   6.035  ops/s
ArrayContainerAndNotRunContainerBenchmark.pairwiseACAndNotRC                    dimension_033  thrpt    6   2783.285 ±  14.780  ops/s
ArrayContainerAndNotRunContainerBenchmark.pairwiseACAndNotRC                     uscensus2000  thrpt    6  26150.184 ±  50.266  ops/s
ArrayContainerAndNotRunContainerBenchmark.pairwiseACAndNotRC                  weather_sept_85  thrpt    6   1596.039 ±  12.919  ops/s
ArrayContainerAndNotRunContainerBenchmark.pairwiseACAndNotRC               wikileaks-noquotes  thrpt    6   3112.493 ±  31.045  ops/s
ArrayContainerAndNotRunContainerBenchmark.pairwiseACAndNotRC                census-income_srt  thrpt    6   2466.304 ±   4.101  ops/s
ArrayContainerAndNotRunContainerBenchmark.pairwiseACAndNotRC                   census1881_srt  thrpt    6  11106.665 ±  20.548  ops/s
ArrayContainerAndNotRunContainerBenchmark.pairwiseACAndNotRC              weather_sept_85_srt  thrpt    6   1351.175 ±   2.567  ops/s
ArrayContainerAndNotRunContainerBenchmark.pairwiseACAndNotRC           wikileaks-noquotes_srt  thrpt    6   8420.691 ±  58.283  ops/s
```

lemire commented:
I'll open an issue regarding off-heap tests.


Regarding your code... That looks good. Do you recommend merging?

lemire commented:
See https://github.com/RoaringBitmap/RoaringBitmap/issues/210

ppiotrow commented:
We can merge it, yes. But there is one more thing that improves the performance. For Mappeable container we can `.clone()` array container at the beginning and then do inplace copies `System.copy(src, ..., src,...,...). That way you can use the code without references to `ShortBuffer` utils. I'll push it to some branch on Monday to get review.

lemire commented:
Ok. I am merging.
